File: samples/getting-started-csharp/Checks.cs
using System.Text.Json;
using Microsoft.FSharp.Core;
using static FCQRS.Common;
using static FCQRS.CSharp;
using static FCQRS.Serialization.Serialization;

public static class DocumentChecks
{
    static void Equal<T>(T expected, T actual, string message)
    {
        if (!Equals(expected, actual)) throw new Exception(message);
    }
    static void Roundtrip<T>(T value) => Equal(value,
        JsonSerializer.Deserialize<T>(JsonSerializer.Serialize(value, jsonOptions), jsonOptions), "Serialized value did not round-trip");
    static Event<DocumentEvent> Event(DocumentEvent value, long version = 1) => TestEnvelope.Event(value, version);
    static Command<DocumentCommand> Command(DocumentCommand value) => TestEnvelope.Command(value);

    public static void Run()
    {
        DocumentRules(); SagaRules(); LegacyRules();
        Console.WriteLine("All document, replay, and saga checks passed.");
    }
    public static void DocumentRules()
    {
        var aggregate = new DocumentAggregate();
        // docs:checks
        var original = new Document("notes", "FCQRS notes", "first event");
        var replacement = original with { Content = "replacement attempt" };
        Equal(EventActions.Persist<DocumentEvent>(new DocumentCreated(original)),
            aggregate.HandleCommand(Command(new CreateDocument(original)), DocumentState.Initial), "First create must store");
        var created = aggregate.ApplyEvent(Event(new DocumentCreated(original)), DocumentState.Initial);
        Equal(EventActions.Defer<DocumentEvent>(new DocumentCreated(original)),
            aggregate.HandleCommand(Command(new CreateDocument(replacement)), created), "Repeated create must preserve the original");
        Equal(created, aggregate.ApplyEvent(Event(new DocumentCreated(original)), created), "Repeated reply must not change state");
        // docs:end
        // docs:edit-checks
        Equal(EventActions.Persist<DocumentEvent>(new DocumentEdited("notes", "second draft")),
            aggregate.HandleCommand(Command(new EditDocument("notes", "second draft")), created), "Edit must store new content");
        var edited = aggregate.ApplyEvent(Event(new DocumentEdited("notes", "second draft"), 2), created);
        Equal(EventActions.Defer<DocumentEvent>(new DocumentEdited("notes", "second draft")),
            aggregate.HandleCommand(Command(new EditDocument("notes", "second draft")), edited), "Repeating the latest edit must defer");
        Equal(edited, aggregate.ApplyEvent(Event(new DocumentEdited("notes", "second draft"), 2), edited), "Deferred edit must not change state");
        Equal(EventActions.Defer<DocumentEvent>(new DocumentRejected("Document does not exist")),
            aggregate.HandleCommand(Command(new EditDocument("missing", "draft")), DocumentState.Initial), "Missing document must reject");
        Equal(edited, aggregate.ApplyEvent(Event(new DocumentRejected("Content must not be blank"), 2), edited), "Deferred rejection must not change state");
        // docs:end
        var waiting = aggregate.ApplyEvent(Event(new PublicationRequested("notes", "guides/fcqrs"), 3), edited);
        var finished = aggregate.ApplyEvent(Event(new PublicationFinished("notes", "guides/fcqrs", PublicationResult.Published), 4), waiting);
        Equal(EventActions.Defer<DocumentEvent>(new PublicationFinished("notes", "guides/fcqrs", PublicationResult.Published)),
            aggregate.HandleCommand(Command(new FinishPublication("guides/fcqrs", PublicationResult.Published)), finished), "Repeated saga completion must defer");
        Equal(EventActions.Defer<DocumentEvent>(new DocumentRejected("Editing closes when publication starts")),
            aggregate.HandleCommand(Command(new EditDocument("notes", "too late")), finished), "Publication freezes editing");
        DocumentEvent[] history = [new DocumentCreated(original), new DocumentEdited("notes", "second draft"),
            new PublicationRequested("notes", "guides/fcqrs"), new PublicationFinished("notes", "guides/fcqrs", PublicationResult.Published)];
        var recovered = history.Select((e, i) => Event(e, i + 1)).Aggregate(DocumentState.Initial, (s, e) => aggregate.ApplyEvent(e, s));
        Equal(finished, recovered, "Mixed-history replay must rebuild the same state");
        foreach (var e in history) Roundtrip(Event(e));
        Roundtrip(Event(new DocumentRejected("reason"))); Roundtrip(finished);
        DocumentCommand[] commands = [new CreateDocument(original), new EditDocument("notes", "draft"),
            new PublishDocument("guides/fcqrs"), new FinishPublication("guides/fcqrs", PublicationResult.Published)];
        foreach (var command in commands) Roundtrip(Command(command));
    }
    static void SagaRules()
    {
        AggregateFactory factory = _ => throw new Exception("Pure tests never resolve actors");
        var saga = new PublicationSaga(factory, factory, false, new());
        SagaState<PublicationData, FSharpOption<PublicationState>> State(PublicationState value) => new(new(), FSharpOption<PublicationState>.Some(value));
        var expiry = new ExpectationExhausted("ReservingSlug", DateTime.UnixEpoch, 3);
        var reserving = new ReservingSlug("notes", "guides/fcqrs");
        var uncertain = new ReservationUncertain("notes", "guides/fcqrs");
        Equal(SagaEventActions.StateChanged<PublicationState>(uncertain), saga.HandleEvent(expiry, State(reserving)), "Exhaustion must remain unknown");
        Equal(SagaEventActions.StateChanged<PublicationState>(new ReportingResult("notes", "guides/fcqrs", PublicationResult.Published)),
            saga.HandleEvent(TestEnvelope.Event<SlugEvent>(new SlugReserved("notes"), 1), State(uncertain)), "Late success must be accepted");
        var reporting = new ReportingResult("notes", "guides/fcqrs", PublicationResult.Published);
        var reportUncertain = new ReportUncertain("notes", "guides/fcqrs", PublicationResult.Published);
        Equal(SagaEventActions.StateChanged<PublicationState>(reportUncertain), saga.HandleEvent(expiry, State(reporting)), "Missing completion reply must remain unknown");
        Equal(SagaEventActions.StateChanged<PublicationState>(new PublicationDone()),
            saga.HandleEvent(Event(new PublicationFinished("notes", "guides/fcqrs", PublicationResult.Published), 4), State(reportUncertain)), "Late completion must finish");
        foreach (PublicationState state in new PublicationState[] { reserving, reporting })
        {
            var effects = saga.ApplySideEffects(new(new(), state), true);
            if (effects.Expect is null || effects.Expect.Resend.Length != 1) throw new Exception("Recovery must re-drive one retry-safe command with a deadline");
        }
        foreach (PublicationState state in new PublicationState[] { reserving, uncertain, reporting, reportUncertain, new PublicationDone() }) Roundtrip(state);
        Roundtrip(TestEnvelope.Event<SlugEvent>(new SlugReserved("notes"), 1));
        Roundtrip(TestEnvelope.Event<SlugEvent>(new SlugUnavailable("other"), 1));
    }
    static void LegacyRules()
    {
        using var fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "fixtures", "document-created-v1-csharp.json")));
        var old = fixture.RootElement.GetProperty("payload").Deserialize<Event<DocumentCreated>>(jsonOptions)!;
        var adapted = (Event<DocumentEvent>)LegacyCreationReader.Read(old);
        Equal(old.Version, adapted.Version, "Legacy reader must preserve the version");
        Equal(old.CorrelationId, adapted.CorrelationId, "Legacy reader must preserve correlation");
        var aggregate = new DocumentAggregate();
        var recovered = aggregate.ApplyEvent(adapted, DocumentState.Initial);
        var edited = aggregate.ApplyEvent(Event(new DocumentEdited(recovered.Document!.Id, "upgraded"), 2), recovered);
        Equal("upgraded", edited.Document!.Content, "Old history must accept a new edit");
        var snapshot = "{\"Document\":" + JsonSerializer.Serialize(old.EventDetails.Document, jsonOptions) + "}";
        Equal(recovered, JsonSerializer.Deserialize<DocumentState>(snapshot, jsonOptions), "Old snapshot with no publication field must recover");
    }
}

File: samples/getting-started-fsharp/Checks.fs
module Checks

open System
open System.IO
open System.Text.Json
open FCQRS.Common
open FCQRS.CSharp
open FCQRS.Serialization.Serialization
open Program

let equal expected actual message = if expected <> actual then failwith message
let command details = TestEnvelope.Command details
let event version details = TestEnvelope.Event(details, version)
let roundtrip<'T when 'T: equality> (value: 'T) =
    equal value (value |> encodeToBytes |> decodeFromBytes<'T>) "Serialized value did not round-trip"

let documentChecks () =
    // docs:checks
    let original = { Id = "notes"; Title = "FCQRS notes"; Content = "first event" }
    let replacement = { original with Content = "replacement attempt" }
    equal (PersistEvent(DocumentCreated original))
        (decide (command (CreateDocument original)) initial) "First create must store"
    let created = fold (event 1L (DocumentCreated original)) initial
    equal (DeferEvent(DocumentCreated original))
        (decide (command (CreateDocument replacement)) created) "Repeated create must preserve the original"
    equal created (fold (event 1L (DocumentCreated original)) created) "Repeated reply must not change state"
    // docs:end
    // docs:edit-checks
    equal (PersistEvent(DocumentEdited("notes", "second draft")))
        (decide (command (EditDocument("notes", "second draft"))) created) "Edit must store new content"
    let edited = fold (event 2L (DocumentEdited("notes", "second draft"))) created
    equal (DeferEvent(DocumentEdited("notes", "second draft")))
        (decide (command (EditDocument("notes", "second draft"))) edited) "Repeating the latest edit must defer"
    equal edited (fold (event 2L (DocumentEdited("notes", "second draft"))) edited) "Deferred edit must not change state"
    equal (DeferEvent(DocumentRejected "Document does not exist"))
        (decide (command (EditDocument("missing", "draft"))) initial) "Missing document must reject"
    equal edited (fold (event 2L (DocumentRejected "Content must not be blank")) edited) "Deferred rejection must not change state"
    // docs:end
    let waiting = fold (event 3L (PublicationRequested("notes", "guides/fcqrs"))) edited
    let finished = fold (event 4L (PublicationFinished("notes", "guides/fcqrs", Published))) waiting
    equal (DeferEvent(PublicationFinished("notes", "guides/fcqrs", Published)))
        (decide (command (FinishPublication("guides/fcqrs", Published))) finished) "Repeated saga completion must defer"
    equal (DeferEvent(DocumentRejected "Editing closes when publication starts"))
        (decide (command (EditDocument("notes", "too late"))) finished) "Publication freezes editing"
    let history = [DocumentCreated original; DocumentEdited("notes", "second draft");
                   PublicationRequested("notes", "guides/fcqrs"); PublicationFinished("notes", "guides/fcqrs", Published)]
    let recovered = history |> List.mapi (fun i e -> event (int64 i + 1L) e) |> List.fold (fun s e -> fold e s) initial
    equal finished recovered "Mixed-history replay must rebuild the same state"
    history |> List.iter (fun e -> roundtrip (event 1L e))
    roundtrip (event 1L (DocumentRejected "reason"))
    roundtrip finished
    [CreateDocument original; EditDocument("notes","draft"); PublishDocument "guides/fcqrs";
     FinishPublication("guides/fcqrs",Published)] |> List.iter (command >> roundtrip)

let sagaChecks () =
    let state value : SagaState<unit, Publication.State option> = { Data = (); State = Some value }
    let expiry = { StateName = "ReservingSlug"; EnteredAt = DateTime.UnixEpoch; Attempts = 3 }
    let reserving = Publication.ReservingSlug("notes", "guides/fcqrs")
    equal (StateChangedEvent(Publication.ReservationUncertain("notes", "guides/fcqrs")))
        (Publication.handleEvent expiry (state reserving)) "Exhaustion must remain unknown"
    let uncertain = Publication.ReservationUncertain("notes", "guides/fcqrs")
    equal (StateChangedEvent(Publication.ReportingResult("notes", "guides/fcqrs", Published)))
        (Publication.handleEvent (event 1L (Publication.Slug.SlugReserved "notes")) (state uncertain)) "Late success must be accepted"
    let reporting = Publication.ReportingResult("notes", "guides/fcqrs", Published)
    equal (StateChangedEvent(Publication.ReportUncertain("notes", "guides/fcqrs", Published)))
        (Publication.handleEvent expiry (state reporting)) "Missing completion reply must remain unknown"
    equal (StateChangedEvent Publication.Done)
        (Publication.handleEvent (event 4L (PublicationFinished("notes", "guides/fcqrs", Published)))
            (state (Publication.ReportUncertain("notes", "guides/fcqrs", Published)))) "Late completion must finish"
    let factory (_: string) : Akkling.Cluster.Sharding.IEntityRef<obj> = failwith "Pure tests never resolve actors"
    for current in [reserving; reporting] do
        let snapshot : SagaState<unit, Publication.State> = { Data = (); State = current }
        let (transition, _) = Publication.applySideEffects factory factory snapshot true
        match transition with
        | StayExpecting expectation -> equal 1 expectation.Resend.Length "Recovery must re-drive one retry-safe command"
        | _ -> failwith "Every active wait must have an expectation"
    [reserving; uncertain; reporting; Publication.ReportUncertain("notes","guides/fcqrs",Published); Publication.Done]
    |> List.iter roundtrip
    [Publication.Slug.SlugReserved "notes"; Publication.Slug.SlugUnavailable "other"] |> List.iter (event 1L >> roundtrip)

let legacyChecks () =
    use fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "fixtures", "document-created-v1-fsharp.json")))
    let root = fixture.RootElement
    let old : Event<DocumentEvent> = decode (root.GetProperty("payload"))
    let recovered = fold old initial
    match recovered.Document with
    | None -> failwith "Old creation did not recover"
    | Some document ->
        let edited = fold (event 2L (DocumentEdited(document.Id, "upgraded"))) recovered
        equal "upgraded" edited.Document.Value.Content "Old history must accept a new edit"
        let snapshot = "{\"Document\":" + encode (Some document) + "}"
        let oldState = decodeFromBytes<DocumentState> (Text.Encoding.UTF8.GetBytes snapshot)
        equal recovered oldState "Old snapshot with no publication field must recover"

let run () =
    documentChecks ()
    sagaChecks ()
    legacyChecks ()
    printfn "All document, replay, and saga checks passed."
